compiler/sem/sanitizer.nim

```diff
@@ -0,0 +1,976 @@
+    checkSonsLen(n, 1)
+    result = stmt(c, n[0])
+  of nkStmtList:
+    result = stmts(c, n)
+  of nkImportStmt, nkImportExceptStmt, nkExportStmt, nkExportExceptStmt, nkFromStmt, nkIncludeStmt:
+    # QUESTION: should we perform any additional validation here or leave that
+    #           to sem?
+    checkMinSonsLen(n, 1)
+    result = exprs(c, n)
+  of nkBindStmt, nkMixinStmt:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i in 0..<n.len:
+      result[i] = qualifiedIdent(c, n[i])
+
+  of nkUsingStmt:
+    # we're much more strict than the parser here: everything besides an ident
+    # def is disallowed
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i in 0..<n.len:
+      result[i] =
+        case n.kind
+        of nkIdentDefs:   strictIdentDefs(c, n[i])
+        of nkCommentStmt: commentStmt(c, n[i])
+        else:             invalidAst(c, n[i])
+
+  of nkCommentStmt:
+    result = commentStmt(c, n)
+  of nkStmtListExpr:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i, it in sliceIt(n, 0..^2):
+      result[i] = stmt(c, it)
+    result[^1] = expr(c, n[^1])
+  of nkBlockExpr:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] = expr(c, n[1])
+  of nkStmtListType:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i, it in sliceIt(n, 0..^2):
+      result[i] = stmt(c, it)
+    result[^1] = typeExpr(c, n[^1])
+  of nkBlockType:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] = typeExpr(c, n[1])
+
+  of nkWith, nkWithout:
+    unreachable("missing")
+
+  of nkTypeOfExpr:
+    checkSonsLen(n, 1)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+  of nkObjectTy,
+    nkTupleTy    ,
+    nkTupleClassTy,
+    nkTypeClassTy  ,
+    nkStaticTy:
+    unreachable("missing")
+  of nkRecList,
+    nkRecCase,
+    nkRecWhen:
+    # context dependent
+    invalid()
+  of nkRefTy,
+    nkPtrTy,
+    nkVarTy,
+    nkConstTy,
+    nkMutableTy,
+    nkDistinctTy,
+    nkProcTy,
+    nkIteratorTy,
+    nkSharedTy,
+    nkEnumTy,
+    nkEnumFieldDef:
+    invalid()
+  of nkArgList:
+    unreachable("missing")
+  of nkPattern:
+    # disallow in macro output
+    invalid()
+  of nkTryStmt, nkHiddenTryStmt:
+    # treat a hidden 'try' statement as a normal one
+    result = parseTry(c, n)
+  of nkClosure:
+    # can't be expressed in untyped AST. Also only present in post-transform
+    # AST
+    # XXX: we could throw away the 'env' part and only use the 'prc' -- I'm
+    #      not sure if that's a good idea however
+    invalid()
+  of nkGotoState, nkState:
+    # only present in post-transform AST. Reject them.
+    invalid()
+  of nkTupleConstr:
+    # can have zero items
+    result = prepareFrom(n)
+    for i, it in n.pairs:
+      result[i] =
+        if it.kind == nkExprColonExpr: parseColonExpr(c, it)
+        else:                          expr(c, it)
+
+  of nkError:
```

Comment:
"technically" this could be:
- if user error, return to `error` pragma
- otherwise, extract the untyped `wrongNode` as that should always be a straight swap

If we enable `error` pragma blocks, then we might have an easier time:
```nim
{.error.}:
  this is not valid syntax without commas
```

Thoughts?

Reply:
I like it.